Hello, my name is Artem. I'm going to review your project!

You can find my comments in <font color='green'>green</font>, <font color='blue'>blue</font> or <font color='red'>red</font> boxes like this:

<div class="alert alert-block alert-success">
<b>Success:</b> if everything is done succesfully
</div>

<div class="alert alert-block alert-info">
<b>Improve: </b> "Improve" comments mean that there are tiny corrections that could help you to make your project better.
</div>

<div class="alert alert-block alert-danger">
<b>Needs fixing:</b> if the block requires some corrections. Work can't be accepted with the red comments.
</div>

### <font color='orange'>General feedback</font>
* You've worked really hard and submitted a solid project.
* Your code style looks good. It was a pleasure to review your project.
* Glad to see that the notebook is well-structured. It's a pleasure to check such projects.
* I've found some tiny mistake in your project. It'll be easy to fix.
* There are also some comments for improvement. They don’t mean that you’ve done anything wrong, but they’re things we wanted to point out that can help you further develop your skills.
* While there's room for improvement, on the whole, your project is impressive/looking good.
* One more time and you'll have it! Good luck!

### <font color='orange'>General feedback (review 2)</font>
* I'm happy to see you've made a few corrections to your work!
* Your project has been accepted and you can go to the next sprint!
* Keep up the good work, and good luck on the next sprint!

## Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building a **credit scoring** of a potential customer. A ** credit scoring ** is used to evaluate the ability of a potential borrower to repay their loan.

### Step 1. Open the data file and have a look at the general information. 

<div class="alert alert-block alert-info">
<b>Improve: </b> Please collect all imports (and nothing else) in the first cell of the project.
</div>

In [26]:
import pandas as pd
from nltk.stem import SnowballStemmer

credit_credentials = pd.read_csv('/datasets/credit_scoring_eng.csv')
credit_credentials.info()
credit_credentials.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

### Conclusion

- In this set of data there are 12 columns. Column names are children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, and purpose. 

- There are 21,525 total rows.

- The column titled days_employed contains negative values,  it would be better to change these to positive values so that the data is easier to read. 

- The columns titled days_employed and total_income contain missing values. 

- The purpose of the loan column will need to be categorized as well because then you can tell how much the loan is for. A car loan will most likely be less in value than a home loan therefore, would not need as much income as a home loan.

<div class="alert alert-block alert-info">
<b>Improve: </b> You could create a bullet point if you've added space after dash (-) in markdown cell above.
</div>

<div class="alert alert-block alert-success">
<b>Success:</b> Data loading and initial analysis were done well.
</div>

### Step 2. Data preprocessing

### Processing missing values

In [27]:
credit_credentials['days_employed'].sort_values()
mean_of_days_employed = credit_credentials['days_employed'][credit_credentials['days_employed'] >0].mean()
median_of_days_employed = credit_credentials['days_employed'][credit_credentials['days_employed'] >0].median()
credit_credentials['days_employed'] = credit_credentials['days_employed'].fillna(mean_of_days_employed)

credit_credentials['total_income'].sort_values()
mean_of_total_income = credit_credentials['total_income'].mean()
median_of_total_income = credit_credentials['total_income'].median()
credit_credentials['total_income'] = credit_credentials['total_income'].fillna(mean_of_total_income)

print('The mean of days_employed is: {:.3f}'.format(mean_of_days_employed))
print('The median of days_employed is: {:.3f}'.format(median_of_days_employed))
print('')
print('The mean of total_income is: {:.3f}'.format(mean_of_total_income))
print('The median of total_income is: {:.3f}'.format(median_of_total_income))

credit_credentials.isna().sum()

The mean of days_employed is: 365004.310
The median of days_employed is: 365213.306

The mean of total_income is: 26787.568
The median of total_income is: 23202.870


children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Conclusion

- The first line of code was to check the values in the days_employed column. Here its discovered that there are negative values mixed in with positive values and NaNs. 

- The second line and third line of code is to find whether or not the missing values in days_employed should be replaced by the mean or median. It also helps to check these because then I can see if there are major outliers in this column that skew this data.

- I did the whole process over for the total_income column and discovered that  there are no major outliers

- All replacements have been made for the columns that contained missing values so then I checked credit_credentials to make sure that there were no more missing values.

- There were the same amount of rows that were missing total_income as well as days_employed perhaps, this was not required to fill out the application?


<div class="alert alert-block alert-success">
<b>Success:</b> It was a possible solution to fill missing values.
    <br> May be total_income and days_employed were optional fields.
</div>

### Data type replacement

In [28]:
credit_credentials['total_income'] = credit_credentials['total_income'].astype('int')
credit_credentials['days_employed'] = credit_credentials['days_employed'].astype('int')

credit_credentials.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


### Conclusion

- Changing the astype to integer was to get rid of the float values

- Float values take up more memory than integers  so by converting to integers the memory the data takes up is decreased



<div class="alert alert-block alert-success">
<b>Success:</b> Well done!
</div>

### Processing duplicates

In [29]:
duplicates = credit_credentials.duplicated()
duplicates_sum = credit_credentials.duplicated().sum()

print(duplicates)
print('')
print(duplicates_sum)
print('')
credit_credentials['family_status'].unique()

0        False
1        False
2        False
3        False
4        False
         ...  
21520    False
21521    False
21522    False
21523    False
21524    False
Length: 21525, dtype: bool

54



array(['married', 'civil partnership', 'widow / widower', 'divorced',
       'unmarried'], dtype=object)

### Conclusion

- The first line of code is to check if there are any duplicates in this table there are 408

- One thing to take into consideration here is that in some columns there are very few options that could be placed there. That is why I pulled information from the family_status column there are only 5 unique values and over 21,000 values. 


<div class="alert alert-block alert-danger">

<b>Needs fixing:</b> You can see at the example below that original row does not marked as duplicated. So in `duplicated` dataframe you could not see 2 first(original) row.

</div>

<div class="alert alert-block alert-danger">
<b>Reviewer code start</b>
</div>

In [5]:
import pandas as pd

df = pd.DataFrame([[1, 2], [1, 3], [1, 2], [1, 2]])
df

0  1
0  1  2
1  1  3
2  1  2
3  1  2

In [7]:
df.duplicated()

0    False
1    False
2     True
3     True
dtype: bool

<div class="alert alert-block alert-danger">
<b>Reviewer code end</b>
</div>

### Categorizing Data

In [30]:
def income_level(income):
    if income <= 20000:
        return 'low income'
    if income <= 50000:
        return 'middle-class'
    if income <= 70000:
        return 'upper-middle-class'
    else:
        return 'upper-class'

credit_credentials['income_level'] = credit_credentials['total_income'].apply(income_level)

def applicant_age(age):
    if age <= 34:
        return 'Young applicant'
    if age <= 49:
        return 'Younger middle age applicant'
    if age <= 65:
        return 'Upper middle age applicant'
    else:
        return 'elder applicant'
credit_credentials['dob_years'] = credit_credentials['dob_years'].apply(applicant_age)

### Conclusion

- I decided to sort the data by income level because it will make it will help to quickly locate which groups of indiviual's data should be reviewed more

- I also decided to group by age this is because currently in the United States the younger you are the more likely you have debt due to university or other outside factors. By grouping by age the reviewer can check to see who is most likely to have other loans and be able to take that into consideration as well.

<div class="alert alert-block alert-success">
<b>Success:</b> This part was done great! Glad to see that "apply" method was used.
</div>

<div class="alert alert-block alert-info">
<b>Improve: </b> You could also make a small analysis. How can it bee -1 or 20 children? Is it possible? What about maximal possible value of days employed?
</div>

### Step 3. Answer these questions

- Is there a relation between having kids and repaying a loan on time?

In [31]:
have_children = credit_credentials['children'][(credit_credentials['children'] !=0)].count()
parents_in_debt = credit_credentials['children'][(credit_credentials['children']!=0) & (credit_credentials['debt']==1)].count()
ratio_of_parental_debt = parents_in_debt / have_children

childless = credit_credentials['children'][(credit_credentials['children']== 0)].count()
childless_debt = credit_credentials['children'][(credit_credentials['children']==0) & (credit_credentials['debt']==1)].count()
ratio_of_childless_debt = childless_debt / childless

children_pivot = credit_credentials.pivot_table(index=['children'], values =['debt'], aggfunc=['sum', 'count'])
children_pivot['% of total'] = children_pivot[('sum', 'debt')] / children_pivot[('count', 'debt')] 
children_pivot['% of total'] = pd.Series(["{0:.2f}%".format(val * 100) for val in children_pivot['% of total']], index = children_pivot.index)
print(children_pivot.sort_values(by=['% of total']))
print('')
print('These amount of applicants have children:', have_children)
print('These amount of applicants that are without children:', childless)
print('')
print('These amount of applicants have children and are in debt:', parents_in_debt)
print('These amount of applicants are without children but, have debt:', childless_debt)
print('')
print('Possility of missed payments with children: {:.2%}'.format(ratio_of_parental_debt))
print('Possibility of missed payments without children: {:.2%}'.format(ratio_of_childless_debt))

           sum  count % of total
          debt   debt           
children                        
 5           0      9      0.00%
 20          8     76     10.53%
-1           1     47      2.13%
 0        1063  14149      7.51%
 3          27    330      8.18%
 1         444   4818      9.22%
 2         194   2055      9.44%
 4           4     41      9.76%

These amount of applicants have children: 7376
These amount of applicants that are without children: 14149

These amount of applicants have children and are in debt: 678
These amount of applicants are without children but, have debt: 1063

Possility of missed payments with children: 9.19%
Possibility of missed payments without children: 7.51%


### Conclusion

- There is a stronger possibility for an applicant to default when they have children

- Interesting to look at is that there are a higher number of indivduals that do not have children but, are in debt

- The more children an applicant has the more likely they are going to miss a payment on their loan.

- Also, it should be noted that there are values that state 20 kids and -1 kids, are these errors in the data?


- Is there a relation between marital status and repaying a loan on time?

In [32]:
martial_pivot = credit_credentials.pivot_table(index=['family_status', 'family_status_id'], values =['debt'], aggfunc=['sum', 'count'])
martial_pivot['% of total'] = martial_pivot[('sum', 'debt')] / martial_pivot[('count', 'debt')] 
martial_pivot['% of total'] = pd.Series(["{0:.2f}%".format(val * 100) for val in martial_pivot['% of total']], index = martial_pivot.index)
print(martial_pivot.sort_values(by=['% of total']))




                                    sum  count % of total
                                   debt   debt           
family_status     family_status_id                       
widow / widower   2                  63    960      6.56%
divorced          3                  85   1195      7.11%
married           0                 931  12380      7.52%
civil partnership 1                 388   4177      9.29%
unmarried         4                 274   2813      9.74%


### Conclusion

- There is a lower probability of missing payments when in a marriage compared to people who are not married. This is significant too becasue there are more people who are married than there are people who are not illustrating that being married is a strong factor to consider when approving loans.

- But, when you break down the not in a marriage people more there is a lot of variance that occurs between these four subcategories.

- Is there a relation between income level and repaying a loan on time?

In [33]:
low_income_total = credit_credentials['income_level'][(credit_credentials['income_level']=='low income')].count()
low_income_debt = credit_credentials['income_level'][(credit_credentials['income_level']=='low income') & (credit_credentials['debt']==1)].count()
ratio_of_low_income_debt = low_income_debt / low_income_total

middle_low_income_total = credit_credentials['income_level'][(credit_credentials['income_level']=='middle-class')].count()
middle_low_income_debt = credit_credentials['income_level'][(credit_credentials['income_level']== 'middle-class') & (credit_credentials['debt']==1)].count()
ratio_of_middle_low_income_debt = middle_low_income_debt / middle_low_income_total

middle_high_income_total = credit_credentials['income_level'][(credit_credentials['income_level']=='upper-middle-class')].count()
middle_high_income_debt = credit_credentials['income_level'][(credit_credentials['income_level']=='upper-middle-class') & (credit_credentials['debt']==1)].count()
ratio_of_middle_high_income_debt = middle_high_income_debt / middle_high_income_total

high_income_total = credit_credentials['income_level'][(credit_credentials['income_level']=='upper-class')].count()
high_income_debt = credit_credentials['income_level'][(credit_credentials['income_level']=='upper-class') & (credit_credentials['debt']==1)].count()
ratio_high_income_debt = high_income_debt / high_income_total


print('Possibility of default with low income: {:.2%}'.format(ratio_of_low_income_debt))
print('Possibility of default with middle low income: {:.2%}'.format(ratio_of_middle_low_income_debt))
print('Possibility of default with middle high income: {:.2%}'.format(ratio_of_middle_high_income_debt))
print('Possibility of default with high income: {:.2%}'.format(ratio_high_income_debt))


Possibility of default with low income: 8.25%
Possibility of default with middle low income: 8.11%
Possibility of default with middle high income: 7.43%
Possibility of default with high income: 5.82%


### Conclusion

- The higher the income the less likely an inidividual will default on their loan payments. 

- I replaced missing values in the total_income column with the mean total_income so this ratio may be incorrect, in order to have the most accurate ratio for this the 2174 values that were replaced need to not be the average but, to be the amount the applicant truly makes

- How do different loan purposes affect on-time repayment of the loan?

In [34]:
english_stemmer = SnowballStemmer('english')


def purpose_stemming(purpose):
    return ''.join([english_stemmer.stem(word)for word in purpose.split()])

credit_credentials['purpose_stemmed']  = credit_credentials['purpose'].apply(purpose_stemming)

def purpose_filter(purpose_stemmed):
    if 'hous' in purpose_stemmed:
        return 'property purpose'
    if 'estat' in purpose_stemmed:
        return 'property purpose'
    if 'properti'in purpose_stemmed:
        return 'property purpose'
    if 'car' in purpose_stemmed:
        return 'vehichle purpose'
    if 'educ' in purpose_stemmed:
        return 'education purpose'
    if 'wed' in purpose_stemmed:
        return 'wedding purpose'
    else:
        return 'other'
credit_credentials['purpose_stemmed']  = credit_credentials['purpose'].apply(purpose_filter)
purpose_pivot = credit_credentials.pivot_table(index=['purpose_stemmed'], values =['debt'], aggfunc=['sum', 'count'])
purpose_pivot['% of total'] = purpose_pivot[('sum', 'debt')] / purpose_pivot[('count', 'debt')] 
purpose_pivot['% of total'] = pd.Series(["{0:.2f}%".format(val * 100) for val in purpose_pivot['% of total']], index = purpose_pivot.index)
print(purpose_pivot.sort_values(by=['% of total']))


                   sum count % of total
                  debt  debt           
purpose_stemmed                        
property purpose   592  8298      7.13%
other              233  3038      7.67%
wedding purpose    186  2348      7.92%
education purpose  327  3526      9.27%
vehichle purpose   403  4315      9.34%


### Conclusion

- A lot of the purpose column contains purposes that are fundamentally the same but, just worded slightly differently

- Vehicle and education loans have a higher probability of defaulting on a loan.



<div class="alert alert-block alert-success">
<b>Success:</b> This step was done perfectly! Glad to see that you've mastered the "pivot_table" method.
</div>

### Step 4. General conclusion

- When looking at this data it is apparent that an applicant who is unmarried, applying for a vehicle loan, with low income, and with children has the <b>highest rate of possibily missing a payment on their loan</b>. 

- In general an applicant that is applying for a home loan who is married is at a <b>low rate for missing a payment</b>. 

- Individuals that are applying for a education loan or a vehicle loan are more likely to default than people who are applying for other reasons. 

- One thing that needs to be considered when looking at this data is that I <b>filled in values</b> that were missing in the <b>total_income</b> value with the mean. Therefore those values may not be accurate and those particular applicants should be reviewed more. 

- I also, created an <b>age filter</b> that could be used to see how age factors into loan repayments and I think that would be smart to utlize as well.

<div class="alert alert-block alert-info">
<b>Improve: </b> It would be better if bullet points, bold font and other markdown opportunities were used to make the conclusion more structured. 
</div>

### Project Readiness Checklist

Put 'x' in the completed points. Then press Shift + Enter.

- [x]  file open;
- [x]  file examined;
- [x]  missing values defined;
- [x]  missing values are filled;
- [x]  an explanation of which missing value types were detected;
- [x]  explanation for the possible causes of missing values;
- [x]  an explanation of how the blanks are filled;
- [x]  replaced the real data type with an integer;
- [x]  an explanation of which method is used to change the data type and why;
- [x]  duplicates deleted;
- [x]  an explanation of which method is used to find and remove duplicates;
- [x]  description of the possible reasons for the appearance of duplicates in the data;
- [x]  data is categorized;
- [x]  an explanation of the principle of data categorization;
- [x]  an answer to the question "Is there a relation between having kids and repaying a loan on time?";
- [X]  an answer to the question " Is there a relation between marital status and repaying a loan on time?";
- [x]   an answer to the question " Is there a relation between income level and repaying a loan on time?";
- [x]  an answer to the question " How do different loan purposes affect on-time repayment of the loan?"
- [x]  conclusions are present on each stage;
- [x]  a general conclusion is made.